Promise variants of form.validate() and field.validate()

voxpelli opened the pull request:
Fixes #217 

Also:

* `async` module is removed and replaced with a tiny tiny helper function
* `field.validate()` now always returns an error, previously it only did so on `required` errors
* due to the fact that `field.validate()` rarely returned an error, the functionality of `validatePastFirstError` was pretty much never respected and the tests was also set up to expect `validatePastFirstError` to not function. So in reality `validatePastFirstError` has been active for most cases for most users, so the restoration of it's functionality can almost be a bit breaking 🤔 Any thoughts @ljharb?


Are the tests enough for `form.validate()`? Its getting late here so I'm running out of time a bit, so I'm opening this PR now anyways so that this code won't be stuck on my computer

To do / decide:

- [ ] How to handle a now functional `validatePastFirstError`
- [ ] Update README with these promise variants of `.validate()` and the one from #215 
- [ ] Once released, do the same, but for the types at DefinitelyTyped, or bring the types in here? (Bringing them here would be easier as then they can be updated prior to release)

voxpelli committed:
[Fix] `validatePastFirstError`: actually respect the option

ljharb commented:
I rebased this. I'd like to remove `async` in its own PR, and tests fail after the first commit, and that's concerning.

voxpelli committed:
[New] fields/forms `.validate()`: add Promise-accepting variant

+ validatePastFirstError is respected
+ field.validate() now always returns an error and that is tested

voxpelli commented:
Rebased it and it passes now

ljharb commented:
It was failing between the two commits, but i tweaked the test plan count and it seems good now.
